File: supabase/functions/_shared/whatsapp.ts
// WhatsApp Cloud API client for Supabase Edge Functions
// Uses environment variables for configuration

const getConfig = () => {
  const accessToken = Deno.env.get("WHATSAPP_ACCESS_TOKEN") ?? "";
  const phoneNumberId = Deno.env.get("WHATSAPP_PHONE_NUMBER_ID") ?? "";
  const apiVersion = Deno.env.get("WHATSAPP_API_VERSION") ?? "v20.0";
  return { accessToken, phoneNumberId, apiVersion };
};

/**
 * Returns true if both WHATSAPP_ACCESS_TOKEN and WHATSAPP_PHONE_NUMBER_ID are set.
 */
export function isConfigured(): boolean {
  const token = Deno.env.get("WHATSAPP_ACCESS_TOKEN");
  const phoneId = Deno.env.get("WHATSAPP_PHONE_NUMBER_ID");
  return !!token && !!phoneId;
}

/**
 * Downloads media from the WhatsApp Cloud API.
 * First retrieves the media URL, then downloads the binary content.
 */
export async function downloadMedia(mediaId: string): Promise<Uint8Array> {
  const { accessToken, apiVersion } = getConfig();

  // Step 1: Get the media URL
  const metadataUrl = `https://graph.facebook.com/${apiVersion}/${mediaId}`;
  const metadataRes = await fetch(metadataUrl, {
    headers: { Authorization: `Bearer ${accessToken}` },
  });

  if (!metadataRes.ok) {
    const errorBody = await metadataRes.text();
    console.error(
      `WhatsApp API error fetching media metadata: ${metadataRes.status}`,
      errorBody,
    );
    throw new Error(
      `Failed to get media URL: ${metadataRes.status} ${errorBody}`,
    );
  }

  const metadata = await metadataRes.json();
  const mediaUrl: string = metadata.url;

  if (!mediaUrl) {
    console.error("WhatsApp API returned no media URL", metadata);
    throw new Error("No media URL returned from WhatsApp API");
  }

  // Step 2: Download the actual media binary
  const mediaRes = await fetch(mediaUrl, {
    headers: { Authorization: `Bearer ${accessToken}` },
  });

  if (!mediaRes.ok) {
    const errorBody = await mediaRes.text();
    console.error(
      `WhatsApp API error downloading media: ${mediaRes.status}`,
      errorBody,
    );
    throw new Error(
      `Failed to download media: ${mediaRes.status} ${errorBody}`,
    );
  }

  const arrayBuffer = await mediaRes.arrayBuffer();
  return new Uint8Array(arrayBuffer);
}

/**
 * Sends a text message via the WhatsApp Cloud API.
 * The `to` param should be digits only (e.g., "919876543210"), no "+" prefix.
 */
export async function sendTextMessage(
  to: string,
  body: string,
): Promise<void> {
  const { accessToken, phoneNumberId, apiVersion } = getConfig();

  const url = `https://graph.facebook.com/${apiVersion}/${phoneNumberId}/messages`;
  const payload = {
    messaging_product: "whatsapp",
    to,
    type: "text",
    text: { body },
  };

  const res = await fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    const errorBody = await res.text();
    console.error(
      `WhatsApp API error sending text message: ${res.status}`,
      errorBody,
    );
    throw new Error(
      `Failed to send text message: ${res.status} ${errorBody}`,
    );
  }
}

/**
 * Sends a template message via the WhatsApp Cloud API.
 * The `to` param should be digits only (e.g., "919876543210"), no "+" prefix.
 */
export async function sendTemplateMessage(
  to: string,
  templateName: string,
  languageCode: string,
  params: string[],
): Promise<void> {
  const { accessToken, phoneNumberId, apiVersion } = getConfig();

  const url = `https://graph.facebook.com/${apiVersion}/${phoneNumberId}/messages`;
  const payload = {
    messaging_product: "whatsapp",
    to,
    type: "template",
    template: {
      name: templateName,
      language: { code: languageCode },
      components: [
        {
          type: "body",
          parameters: params.map((p) => ({ type: "text", text: p })),
        },
      ],
    },
  };

  const res = await fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    const errorBody = await res.text();
    console.error(
      `WhatsApp API error sending template message: ${res.status}`,
      errorBody,
    );
    throw new Error(
      `Failed to send template message: ${res.status} ${errorBody}`,
    );
  }
}

File: backend/app/infrastructure/whatsapp.py
"""WhatsApp messaging — routes through WaCRM when configured, else direct Meta Cloud API.

When WACRM_API_URL + WACRM_API_KEY are set, all sends go through WaCRM's public API,
giving TurboFix shared inbox, contact management, broadcasts, and conversation tracking.
Falls back to direct Meta Graph API calls when WaCRM is not configured.
"""

import mimetypes
from typing import List

from app import config
from app.infrastructure.http_client import resilient_get, resilient_post
from app.infrastructure.logging import get_logger

log = get_logger("turbofix.whatsapp")


def _use_wacrm() -> bool:
    from app.infrastructure.wacrm_client import is_configured
    return is_configured()


def _graph_url(path: str) -> str:
    return f"https://graph.facebook.com/{config.WHATSAPP_API_VERSION}/{path}"


async def download_media(media_id: str) -> str:
    """Resolve a WhatsApp media ID → download → save to MEDIA_STORE_DIR.

    Returns the local file path as a string (consumed by the AI transcription step).
    Raises on any HTTP/network error after retries.
    """
    headers = {"Authorization": f"Bearer {config.WHATSAPP_ACCESS_TOKEN}"}

    meta_resp = await resilient_get(_graph_url(media_id), headers=headers)
    meta = meta_resp.json()

    media_resp = await resilient_get(meta["url"], headers=headers)

    mime_type = meta.get("mime_type", "application/octet-stream")
    extension = mimetypes.guess_extension(mime_type.split(";")[0].strip()) or ""
    dest = config.MEDIA_STORE_DIR / f"{media_id}{extension}"
    dest.write_bytes(media_resp.content)
    log.info("whatsapp.media_downloaded", media_id=media_id, path=str(dest))
    return str(dest)


async def send_template_message(to: str, params: List[str]) -> None:
    """Send the pre-approved ticket notification template to `to`."""
    if _use_wacrm():
        from app.infrastructure import wacrm_client
        await wacrm_client.send_template_message(
            to=to,
            template_name=config.WHATSAPP_TICKET_TEMPLATE_NAME,
            language=config.WHATSAPP_TICKET_TEMPLATE_LANGUAGE,
            params=params,
        )
        return

    headers = {
        "Authorization": f"Bearer {config.WHATSAPP_ACCESS_TOKEN}",
        "Content-Type": "application/json",
    }
    payload = {
        "messaging_product": "whatsapp",
        "to": to,
        "type": "template",
        "template": {
            "name": config.WHATSAPP_TICKET_TEMPLATE_NAME,
            "language": {"code": config.WHATSAPP_TICKET_TEMPLATE_LANGUAGE},
            "components": [{
                "type": "body",
                "parameters": [{"type": "text", "text": str(p)} for p in params],
            }],
        },
    }
    await resilient_post(
        _graph_url(f"{config.WHATSAPP_PHONE_NUMBER_ID}/messages"),
        headers=headers,
        json=payload,
    )
    log.info("whatsapp.template_sent", to=to, template=config.WHATSAPP_TICKET_TEMPLATE_NAME)


async def send_closure_template(to: str, params: List[str]) -> None:
    """Send the ticket-closed notification template to `to`."""
    if _use_wacrm():
        from app.infrastructure import wacrm_client
        await wacrm_client.send_template_message(
            to=to,
            template_name=config.WHATSAPP_CLOSURE_TEMPLATE_NAME,
            language=config.WHATSAPP_CLOSURE_TEMPLATE_LANGUAGE,
            params=params,
        )
        return

    headers = {
        "Authorization": f"Bearer {config.WHATSAPP_ACCESS_TOKEN}",
        "Content-Type": "application/json",
    }
    payload = {
        "messaging_product": "whatsapp",
        "to": to,
        "type": "template",
        "template": {
            "name": config.WHATSAPP_CLOSURE_TEMPLATE_NAME,
            "language": {"code": config.WHATSAPP_CLOSURE_TEMPLATE_LANGUAGE},
            "components": [{
                "type": "body",
                "parameters": [{"type": "text", "text": str(p)} for p in params],
            }],
        },
    }
    await resilient_post(
        _graph_url(f"{config.WHATSAPP_PHONE_NUMBER_ID}/messages"),
        headers=headers,
        json=payload,
    )
    log.info("whatsapp.closure_sent", to=to, template=config.WHATSAPP_CLOSURE_TEMPLATE_NAME)


async def send_text_message(to: str, text: str) -> None:
    """Send a plain text message."""
    if _use_wacrm():
        from app.infrastructure import wacrm_client
        await wacrm_client.send_text_message(to, text)
        return

    headers = {
        "Authorization": f"Bearer {config.WHATSAPP_ACCESS_TOKEN}",
        "Content-Type": "application/json",
    }
    payload = {
        "messaging_product": "whatsapp",
        "to": to,
        "type": "text",
        "text": {"body": text},
    }
    await resilient_post(
        _graph_url(f"{config.WHATSAPP_PHONE_NUMBER_ID}/messages"),
        headers=headers,
        json=payload,
    )
    log.info("whatsapp.text_sent", to=to)


async def send_escalation_template(to: str, params: List[str]) -> None:
    """Send escalation notification template."""
    if _use_wacrm():
        from app.infrastructure import wacrm_client
        await wacrm_client.send_template_message(
            to=to,
            template_name=config.WHATSAPP_ESCALATION_TEMPLATE_NAME,
            language=config.WHATSAPP_ESCALATION_TEMPLATE_LANGUAGE,
            params=params,
        )
        return

    headers = {
        "Authorization": f"Bearer {config.WHATSAPP_ACCESS_TOKEN}",
        "Content-Type": "application/json",
    }
    payload = {
        "messaging_product": "whatsapp",
        "to": to,
        "type": "template",
        "template": {
            "name": config.WHATSAPP_ESCALATION_TEMPLATE_NAME,
            "language": {"code": config.WHATSAPP_ESCALATION_TEMPLATE_LANGUAGE},
            "components": [{
                "type": "body",
                "parameters": [{"type": "text", "text": str(p)} for p in params],
            }],
        },
    }
    await resilient_post(
        _graph_url(f"{config.WHATSAPP_PHONE_NUMBER_ID}/messages"),
        headers=headers,
        json=payload,
    )
    log.info("whatsapp.escalation_sent", to=to)


async def send_broadcast(name: str, template_name: str, language: str,
                         recipients: List[dict]) -> dict:
    """Send a broadcast to multiple recipients. Only available via WaCRM.

    recipients: [{"to": "+phone", "params": ["var1"]}, ...]
    Falls back to individual sends if WaCRM is not configured.
    """
    if _use_wacrm():
        from app.infrastructure import wacrm_client
        return await wacrm_client.launch_broadcast(
            name=name,
            template_name=template_name,
            template_language=language,
            recipients=recipients,
        )

    sent = 0
    for r in recipients:
        try:
            to = r.get("to", "")
            params = r.get("params", [])
            headers = {
                "Authorization": f"Bearer {config.WHATSAPP_ACCESS_TOKEN}",
                "Content-Type": "application/json",
            }
            payload = {
                "messaging_product": "whatsapp",
                "to": to,
                "type": "template",
                "template": {
                    "name": template_name,
                    "language": {"code": language},
                    "components": [{
                        "type": "body",
                        "parameters": [{"type": "text", "text": str(p)} for p in params],
                    }],
                },
            }
            await resilient_post(
                _graph_url(f"{config.WHATSAPP_PHONE_NUMBER_ID}/messages"),
                headers=headers,
                json=payload,
            )
            sent += 1
        except Exception as exc:
            log.error("whatsapp.broadcast_individual_failed", to=r.get("to"), error=str(exc))

    return {"total_recipients": len(recipients), "accepted": sent}
